### PR TITLE
Serialize last valid state for future change comparison

### PR DESCRIPTION
**Pull request by nrobinaubertin**
### Overview

The zone-transfer mechanism doesn't copy the "ORIGIN" line from the zone file to the secondaries and it was used to store the hash of the zone to check for changes before reloading bind.  
Adding to that, parsing the config files to get the network topology of the last valid state was error prone and we also couldn't detect the removal of a zone.

We now serialize the last valid state (composed of the network topology and the zones) into a json file and load it to check for changes before recomputing the configuration and reloading bind.  
This ends up removing the code related to parsing the metadata and IPs from bind configuration files.

The change in our any-charm usage was made to allow for after deployment changes in the requested dns records (the dns entries file cannot be defined during deployment: it gets overriden by any-charm while handling later events).

I also had to add `also-notify` to the zone config to effectively propagate the zone change event and initiate the zone transfer. An alternative would be to use the `NS` record but I decided against it in the interest of simplicity for the time being.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/dns_data.py [hunk 1]
```diff
@@ -64,50 +50,36 @@ def create_dns_record_provider_data(
...
-    if topology is not None and topology.is_current_unit_active:
-        try:
-            named_conf_content = pathlib.Path(
-                constants.DNS_CONFIG_DIR, "named.conf.local"
-            ).read_text(encoding="utf-8")
-        except FileNotFoundError:
+    if "zones" not in last_valid_state or last_valid_state["zones"] != zones:
+        return True
+
+    if topology is not None and "topology" in last_valid_state:
```
Comment (arturo-seijas) [inline src/dns_data.py:78]: I believe you are missing
```
if topology is None and "topology" in last_valid_state:
    return True
```

File: tests/integration/helpers.py [hunk 2]
```diff
@@ -161,10 +161,6 @@ async def generate_anycharm_relation(
     any_charm_src_overwrite = {
         "any_charm.py": any_charm_content,
         "dns_record.py": dns_record_content,
-        # It's okay to write to /tmp for these tests, so # nosec is used
```
Comment (arturo-seijas) [inline tests/integration/helpers.py:164]: Why this change?

Comment (nrobinaubertin) [reply tests/integration/helpers.py:164]: I was trying to understand why the file was overwritten every time I tried to edit it when the charm was already deployed. The location has nothing to do with it but since `/srv` doesn't trigger the static analysis warnings, I kept it

File: src/dns_data.py [hunk 3]
```diff
@@ -64,50 +50,36 @@ def create_dns_record_provider_data(
...
+        if last_valid_state["topology"] is None and topology is not None:
```
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply src/dns_data.py]: Fixed

File: src/dns_data.py (hunk 1 above)
Comment (nrobinaubertin) [reply src/dns_data.py]: I revisited that and reworked it to be way simpler.

**Review by arturo-seijas — APPROVED** [review]

File: src/bind.py [hunk 4]
```diff
@@ -293,21 +298,20 @@ def _generate_named_conf_local(
...
-                    absolute_path=f"{constants.DNS_CONFIG_DIR}/db.{name}",
-                    zone_transfer_ips=self._bind_config_ip_list(topology.standby_units_ip),
-                )
-            else:
-                content += constants.NAMED_CONF_SECONDARY_ZONE_DEF_TEMPLATE.format(
-                    name=name,
-                    absolute_path=f"{constants.DNS_CONFIG_DIR}/db.{name}",
-                    primary_ip=self._bind_config_ip_list([topology.active_unit_ip]),
-                )
+        if topology is not None:
```
Comment (yanksyoon) [inline src/bind.py:301]: Nitpick:
I think if we reverse the return order 
```suggestion
if topology is None:
    return content
...
```
it might help reduce 1 indentation :D

File: src/dns_data.py [hunk 5]
```diff
@@ -64,50 +50,36 @@ def create_dns_record_provider_data(
...
-        )
+
+    # We assume "zones" and "topology" keys exist in last_valid_state
+    # As it was created from load_state()
+
+    if zones != last_valid_state["zones"]:
+        return True
+
+    if topology != last_valid_state["topology"]:
+        return True
```
Comment (yanksyoon) [inline src/dns_data.py:82]: A very nitpicky comment - i think your approach reads better though but i'll leave the thought.

```suggestion
    return zones != last_valid_state["zones"] or topology != last_valid_state["topology"]
```

File: tests/integration/helpers.py [hunk 6]
```diff
@@ -109,11 +109,11 @@ async def push_to_unit(
     temp_filename_on_workload = _generate_random_filename()
     # unit does have scp_to
     await unit.scp_to(source=temp_path, destination=temp_filename_on_workload)  # type: ignore
-    mv_cmd = f"mv /home/ubuntu/{temp_filename_on_workload} {destination}"
+    mv_cmd = f"sudo mv -f /home/ubuntu/{temp_filename_on_workload} {destination}"
```
Comment (yanksyoon) [inline tests/integration/helpers.py:112]: Just out of curiosity what is the -f for?

**Review by yanksyoon — APPROVED** [review]
LGTM! Thanks for walking me through the PR today!

File: tests/integration/helpers.py (hunk 6 above)
Comment (nrobinaubertin) [reply tests/integration/helpers.py:112]: This is to force the file replacement without prompting to confirm

File: src/dns_data.py (hunk 5 above)
Comment (nrobinaubertin) [reply src/dns_data.py:82]: I actually like yours better, I'll change that in an upcoming PR
